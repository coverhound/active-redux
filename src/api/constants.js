export const API_CONFIGURE = 'AR_API_CONFIGURE';
export const API_CLEAR = 'AR_API_CLEAR';
export const API_HYDRATE = 'API_HYDRATE';

export const API_WILL_READ = 'AR_API_WILL_READ';
export const API_READ_DONE = 'AR_API_READ_DONE';
export const API_READ_FAILED = 'AR_API_READ_FAILED';

export const API_WILL_CREATE = 'AR_API_WILL_CREATE';
export const API_CREATE_DONE = 'AR_API_CREATE_DONE';
export const API_CREATE_FAILED = 'AR_API_CREATE_FAILED';

export const API_WILL_UPDATE = 'AR_API_WILL_UPDATE';
export const API_UPDATE_DONE = 'AR_API_UPDATE_DONE';
export const API_UPDATE_FAILED = 'AR_API_UPDATE_FAILED';

export const API_WILL_DELETE = 'AR_API_WILL_DELETE';
export const API_DELETE_DONE = 'AR_API_DELETE_DONE';
export const API_DELETE_FAILED = 'AR_API_DELETE_FAILED';

export const API_WILL_INDEX = 'AR_API_WILL_INDEX';
export const API_INDEX_DONE = 'AR_API_INDEX_DONE';
export const API_INDEX_CLEAR = 'AR_API_INDEX_CLEAR';
