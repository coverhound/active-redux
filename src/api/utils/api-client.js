import axios from 'axios';
import createError from 'axios/lib/core/createError';

const jsonContentTypes = [
  'application/json',
  'application/vnd.api+json',
];

const validateContentType = (response) => {
  const contentType = response.headers['content-type'].split(';')[0];
  if (jsonContentTypes.includes(contentType)) return;

  throw createError(
    'Invalid Content-Type in response',
    response.config,
    null,
    response
  );
};

const mergeDefaults = (url, options) => ({
  ...options,
  url,
  headers: {
    Accept: 'application/vnd.api+json',
    'Content-Type': 'application/vnd.api+json',
    ...options.headers,
  },
});

export default (url, options = {}) => (
  axios(mergeDefaults(url, options)).then((response) => {
    if (response.status === 204) return response;
    validateContentType(response);
    return response.data;
  })
);
