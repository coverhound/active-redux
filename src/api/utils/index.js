import imm from 'object-path-immutable';
import Registry from '../../registry';
import '../../polyfill';

export apiClient from './api-client';

/**
 * @module
 * @private
 */

/**
 * @example
 * incrementProperty(state, 'isReading')
 * @param {Object} state Object on which the key is incremented
 * @param {string} key Key to increment
 */
export const incrementProperty = (state, key) => (
  imm.set(state, key, state[key] + 1)
);

/**
 * @example
 * decrementProperty(state, 'isReading')
 * @param {Object} state Object on which the key is decremented
 * @param {string} key Key to decrement
 */
export const decrementProperty = (state, key) => (
  imm.set(state, key, state[key] - 1)
);

/**
 * @param {Array|Object} data
 */
export const resourcesArray = (data = []) => (
  Array.isArray(data) ? data : [data]
);

/**
 * @example
 * const obj = { foo: bar: { baz: 'hi' } }
 * getProperty(obj, 'foo', 'bar', 'baz') // => 'hi'
 * getProperty(obj, 'a', 'b', 'c') // => undefined
 *
 * @param {Object} object Object to traverse
 * @param {String[]} keys Keys to look up
 */
export const getProperty = (object, ...keys) => (
  keys.reduce((acc, key) => {
    if (acc === undefined) return acc;
    return acc[key];
  }, object)
);

/**
 * Creates reverse relationships for data merged into the state
 * @param {Object} state
 * @param {Object} newState Immutable Object via object-path-immutable
 * @param {Object} data JSON-API data object
 * @param {Object} data.id
 * @param {Object} data.type
 * @param {Object} data.relationships
 */
export const createReverseRelationships = (state, newState, { id, type, relationships = {} }) => {
  Object.values(relationships).forEach(({ data }) => {
    if (!data) return;

    const children = resourcesArray(data);
    const childModel = Registry.get(children[0].type);
    const childData = { type, id };
    if (!childModel.relationships[type]) return;

    const { key, isArray } = childModel.relationships[type];

    children.forEach(({ type: relType, id: relId }) => {
      if (!getProperty(state, 'resources', relType, relId)) return;
      const objectPath = ['resources', relType, relId, 'relationships', key];
      const existingRelationships = resourcesArray(getProperty(state, ...objectPath));

      if (!isArray) {
        newState.set([...objectPath, 'data'], childData);
        return;
      }

      if (existingRelationships.find((relationship) => relationship.id === id)) return;

      newState.push([...objectPath, 'data'], childData);
    });
  });
};

/**
 * Merges all resources into the state
 * @param {Object} state
 * @param {Object} payload JSON-API payload
 * @param {Object} payload.data
 * @param {Object} payload.included
 */
export const mergeResources = (state, { data, included = [] }) => {
  const newState = imm(state);

  resourcesArray(data).concat(included).forEach((dataObj) => {
    const { id, type } = dataObj;
    newState.set(['resources', type, String(id)], dataObj);
    createReverseRelationships(state, newState, dataObj);
  });

  return newState.value();
};

/**
 * This function clears differently based on what's passed in
 * Object: { type, id } -> this resource is cleared
 * Array: Each of the resources is cleared
 * String: '[type]' -> These resources are cleared
 * undefined: -> All resources are cleared
 * @param {Object} state
 * @param {Object[]|Object|string|undefined} data
 */
export const clearResources = (state, data) => {
  switch (typeof data) {
    case 'object': {
      if (Array.isArray(data)) return data.reduce(clearResources, state);
      const { type, id } = data;
      return imm.del(state, ['resources', type, id]);
    }
    case 'string': return imm.set(state, ['resources', data], {});
    case 'undefined': return imm.set(state, 'resources', {});
    default: return state;
  }
};

/**
 * @example
 * const CUSTOM_REDUX_ACTION = 'CUSTOM_REDUX_ACTION';
 * export const customReduxAction = createAction(CUSTOM_REDUX_ACTION); // => [function]
 * customReduxAction(payload) // => { type, payload }
 * @param {string} type Action type
 */
export const createAction = (type) => (payload) => ({ type, payload });

/**
 * @typedef {Object} ReducerFactory
 * @property {Object} map Action to function mapping
 * @property {Object} initialState Initial reducer state
 *
 * @param {ReducerFactory} base
 * @param {...ReducerFactory} extensions
 */
export const createReducer = ({ map, initialState }, ...extensions) => {
  const extend = (...addedExtensions) => (
    createReducer({ map, initialState }, ...extensions, ...addedExtensions)
  );

  const reducerFactory = imm({ map, initialState }).value();
  extensions.forEach(({ map: m, initialState: i }) => {
    Object.assign(reducerFactory.map, m);
    Object.assign(reducerFactory.initialState, i);
  });

  const reducer = (state = reducerFactory.initialState, action) => {
    const reduce = reducerFactory.map[action.type];
    return reduce === undefined ? state : reduce(state, action);
  };

  reducer.extend = extend;

  return reducer;
};

/**
 * @param {Object} state
 * @param {Object} resource
 * @param {Array|Object} resource.data
 */
export const markPendingResources = (state, { data }) => (
  resourcesArray(data).reduce((acc, resource) => {
    if (!getProperty(state, 'resources', resource.type, resource.id)) return acc;
    return acc.set(['resources', resource.type, resource.id, 'isPending'], true);
  }, imm(state)).value()
);
