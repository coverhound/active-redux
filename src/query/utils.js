import qs from 'qs';
import { createSelectorCreator, defaultMemoize } from 'reselect';

const isEqual = (a, b) => (
  typeof a === 'string' ?
    a === b :
    JSON.stringify(a) === JSON.stringify(b)
);

const createDeepEqualSelector = createSelectorCreator(
  defaultMemoize,
  isEqual,
);

/**
 * @alias module:active-redux.createQuerySelector
 * @function
 * @example
 *
 * import { createQuerySelector } from 'active-redux';
 * import { Comment } from '@models';
 *
 * const Comment = ({ comment }) => <p>{comment.body}</p>;
 *
 * // needed in order to have a unique selector per instance of the component
 * const mapStateToProps = () => {
 *   const querySelector = createQuerySelector(Comment, 'find');
 *   return (state, props) => ({
 *     // Arguments passed into querySelector are passed to `Comment#find`
 *     // and turned into a selector
 *     comments: querySelector(props.commentId)(state)
 *   });
 * };
 *
 * const mapDispatchToProps = () => ({});
 *
 * export default connect(mapStateToProps, mapDispatchToProps)(Comment);
 */
export const createQuerySelector = (promise, selector) => createDeepEqualSelector(
  (...args) => args,
  (args) => {
    promise(...args);
    const out = selector(...args);
    out.promise = promise;
    return out;
  }
);

export const generateEndpoint = (endpoint, query) => {
  let queryString = query;
  if (typeof query !== 'string') {
    queryString = qs.stringify(query);
  }
  return `${endpoint}?${queryString}`;
};
