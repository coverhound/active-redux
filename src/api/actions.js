import * as ActionTypes from './constants';
import { createAction, apiClient } from './utils';
/**
 * @module active-redux/api
 */

const wrappedApiRequest = ({
  resource,
  options,
  endpoint,
  dispatch,
  success,
  failure,
}) => apiClient(endpoint, options).then((json) => {
  dispatch(success(json));
  return Promise.resolve(json);
}).catch((error) => {
  const err = error;
  err.resource = resource;

  dispatch(failure(err));
  return Promise.reject(err);
});

/**
 * Configure the API
 * @see {@link https://github.com/mzabriskie/axios#request-config}
 * @function
 * @param {Object} config - Axios configuration
 */
export const apiConfigure = createAction(ActionTypes.API_CONFIGURE);

/**
 * Clear an entity from the store
 * @function
 * @private
 * @example
 * import { api } from 'active-redux';
 *
 * state.api.people
 * // => Object<String: Person>
 *
 * dispatch(api.apiClear('people'))
 *
 * state.api.people
 * // => {}
 * @param {Object} data - JSON-API data
 */
export const apiClear = createAction(ActionTypes.API_CLEAR);

/**
 * Hydrate the store from JSON-API
 * @function
 * @example
 * import { api } from 'active-redux';
 *
 * state.api.people
 * // => {}
 *
 * const data = [{
 *   type: 'people',
 *   id: '5',
 *   attributes: {
 *     name: 'Joe',
 *     age: 35,
 *   }
 * }];
 *
 * dispatch(api.apiHydrate(data))
 *
 * state.api.people
 * // => Object<String: Person>
 * @param {Object} data - JSON-API data
 */
export const apiHydrate = createAction(ActionTypes.API_HYDRATE);

const apiWillCreate = createAction(ActionTypes.API_WILL_CREATE);
const apiCreateDone = createAction(ActionTypes.API_CREATE_DONE);
const apiCreateFailed = createAction(ActionTypes.API_CREATE_FAILED);

/**
 * Create a resource via API
 * @function
 * @example
 * import { api } from 'active-redux';
 * import Person from '../models/person';
 *
 * Person.find({ id: '5' });
 * // => Promise<null>
 *
 * const person = new Person({ attributes: { name: 'Joe' } });
 *
 * dispatch(api.apiCreate({ resource: person })).then((json) => {
 *   // do something with the response json
 * }).catch((error) => {
 *   // handle error
 * });
 *
 * Person.find({ id: '5' });
 * // => Promise<Person>
 * @param {Object} args
 * @param {define~Model} args.resource
 * @param {string} [args.endpoint]
 */
export const apiCreate = ({ resource, endpoint = resource.endpoint('create') }) => (dispatch, getState) => {
  dispatch(apiWillCreate(resource));
  return wrappedApiRequest({
    resource,
    options: {
      ...getState().api.apiConfig,
      method: 'POST',
      data: { data: resource }
    },
    endpoint,
    dispatch,
    success: apiCreateDone,
    failure: apiCreateFailed,
  });
};

const apiWillRead = createAction(ActionTypes.API_WILL_READ);
const apiReadDone = createAction(ActionTypes.API_READ_DONE);
const apiReadFailed = createAction(ActionTypes.API_READ_FAILED);

/**
 * Read a resource via API
 * @function
 * @example
 * import { api } from 'active-redux';
 * import Person from '../models/person';
 *
 * Person.all();
 * // => Promise<[]>
 *
 * dispatch(api.apiRead({ resource: Person })).then((json) => {
 *   // do something with the response json
 * }).catch((error) => {
 *   // handle error
 * });
 *
 * Person.all();
 * // => Promise<Array<Person>>
 * @param {Object} args
 * @param {Model} args.resource
 * @param {string} [args.endpoint]
 */
export const apiRead = ({ resource, endpoint = resource.endpoint('read') }) => (dispatch, getState) => {
  dispatch(apiWillRead(resource));
  return wrappedApiRequest({
    resource,
    options: {
      ...getState().api.apiConfig,
      method: 'GET',
      data: { data: resource }
    },
    endpoint,
    dispatch,
    success: apiReadDone,
    failure: apiReadFailed,
  });
};

const apiWillUpdate = createAction(ActionTypes.API_WILL_UPDATE);
const apiUpdateDone = createAction(ActionTypes.API_UPDATE_DONE);
const apiUpdateFailed = createAction(ActionTypes.API_UPDATE_FAILED);

/**
 * Update a resource via API
 * @function
 * @example
 * import { api } from 'active-redux';
 * import Person from '../models/person';
 *
 * Person.find({ id: '5' });
 * // => Promise<Person(id='5' name='Joe')>
 *
 * const person = new Person({ id: '5', attributes: { name: 'Jimmy' } });
 *
 * dispatch(api.apiUpdate({ resource: person })).then((json) => {
 *   // do something with the response json
 * }).catch((error) => {
 *   // handle error
 * });
 *
 * Person.find({ id: '5' });
 * // => Promise<Person(id='5' name='Jimmy')>
 * @param {Object} args
 * @param {Model} args.resource
 * @param {string} [args.endpoint]
 */
export const apiUpdate = ({ resource, endpoint = resource.endpoint('update') }) => (
  (dispatch, getState) => {
    dispatch(apiWillUpdate(resource));
    return wrappedApiRequest({
      resource,
      options: {
        ...getState().api.apiConfig,
        method: 'PATCH',
        data: { data: resource }
      },
      endpoint,
      dispatch,
      success: apiUpdateDone,
      failure: apiUpdateFailed,
    });
  }
);

const apiWillDelete = createAction(ActionTypes.API_WILL_DELETE);
const apiDeleteDone = createAction(ActionTypes.API_DELETE_DONE);
const apiDeleteFailed = createAction(ActionTypes.API_DELETE_FAILED);

/**
 * Delete a resource via API
 * @function
 * @example
 * import { api } from 'active-redux';
 * import Person from '../models/person';
 *
 * Person.find({ id: '5' });
 * // => Promise<Person(id='5')>
 *
 * const person = new Person({ id: '5' });
 *
 * dispatch(api.apiDelete({ resource: person })).then((json) => {
 *   // do something with the response json
 * }).catch((error) => {
 *   // handle error
 * });
 *
 * Person.find({ id: '5' });
 * // => Promise<null>
 * @param {Object} args
 * @param {Model} args.resource
 * @param {string} [args.endpoint]
 */
export const apiDelete = ({ resource, endpoint = resource.endpoint('delete') }) => (
  (dispatch, getState) => {
    const options = {
      ...getState().api.apiConfig,
      method: 'DELETE',
    };

    dispatch(apiWillDelete(resource));
    return apiClient(endpoint, options).then((json) => {
      dispatch(apiDeleteDone(resource));
      return Promise.resolve(json);
    }).catch((error) => {
      const err = error;
      err.resource = resource;

      dispatch(apiDeleteFailed(err));
      return Promise.reject(err);
    });
  }
);

const apiWillIndex = createAction(ActionTypes.API_WILL_INDEX);
const apiIndexDone = createAction(ActionTypes.API_INDEX_DONE);
export const apiIndexClear = createAction(ActionTypes.API_INDEX_CLEAR);

export const apiIndexAsync = ({ hash, promise }) => (dispatch, getState) => {
  const indexExists = () => getState().api.indices[hash];

  if (indexExists() && indexExists().isFetching) return promise;

  dispatch(apiWillIndex(hash));

  return promise.then((resources) => {
    if (!indexExists()) return resources;
    dispatch(apiIndexDone({ hash, resources }));
    return resources;
  });
};
export const apiIndexSync = ({ hash, resources }) => (dispatch) => {
  dispatch(apiWillIndex(hash));
  dispatch(apiIndexDone({ hash, resources }));
};
