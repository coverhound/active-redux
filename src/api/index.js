import { createReducer } from './utils';
import baseReducer from './reducer';

export * from './reducer';

/**
 * API Reducer
 * @function
 * @alias module:active-redux/api
 * @example
 * import { createStore, combineReducers, applyMiddleware } from 'redux';
 * import { api } from 'active-redux';
 *
 * const initialState = {};
 *
 * export default createStore(
 *   combineReducers({ api: api.reducer }),
 *   initialState,
 *   applyMiddleware(thunk),
 * );
 *
 * @example
 * import { createStore, combineReducers, applyMiddleware } from 'redux';
 * import { api } from 'active-redux';
 *
 * const initialState = {};
 *
 * // You can add your own extensions to the reducer
 * const customExtensions = [{
 *   map: { AR_PAGINATE_MODELS: (state, payload) => {} },
 *   initialState: { pagination: {} },
 * }];
 *
 * export default createStore(
 *   combineReducers({ api: api.reducer.extend(...customExtension) }),
 *   initialState,
 *   applyMiddleware(thunk),
 * );
 * @param {Object} state
 * @param {Object} action
 */
export const reducer = createReducer(baseReducer);
