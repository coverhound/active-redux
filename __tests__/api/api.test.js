import { api as subject } from 'active-redux';
import * as ActionTypes from 'active-redux/api/constants';
import { createAction } from 'active-redux/api/utils';
import mockStore from 'fixtures/store';

describe('reducer', () => {
  const person = {
    type: 'people',
    id: 5,
    attributes: {
      hair: 'black',
      age: 25,
    }
  };

  const fullStore = {
    apiConfig: {},
    indices: {},
    isCreating: 0,
    isDeleting: 0,
    isReading: 0,
    isUpdating: 0,
    resources: {
      [person.type]: {
        [person.id]: { attributes: { name: 'Johnny' } },
        2: { attributes: { name: 'Yangster' } }
      }
    }
  };

  describe('initialState', () => {
    it('has an empty resources object', () => {
      expect(subject.reducer(undefined, {})).toMatchSnapshot();
    });
  });

  describe('actions', () => {
    const reduce = subject.reducer;

    describe('apiConfigure()', () => {
      it('updates the apiConfig key', () => {
        const state = reduce(undefined, subject.apiConfigure({ baseUrl: 'http://coverhound.com/api' }));
        expect(state).toMatchSnapshot();
      });
    });


    describe('apiHydrate()', () => {
      it('populates the state with resources', () => {
        const state = reduce(undefined, subject.apiHydrate({ data: person }));
        expect(state).toMatchSnapshot();
      });
    });

    describe('apiCreate()', () => {
      let state = fullStore;
      const apiWillCreate = createAction(ActionTypes.API_WILL_CREATE);
      const apiCreateDone = createAction(ActionTypes.API_CREATE_DONE);

      it('increments isCreating', () => {
        state = reduce(undefined, apiWillCreate({ data: person }));
        expect(state).toMatchSnapshot();
      });

      it('populates the state with resources', () => {
        state = reduce(state, apiCreateDone({ data: person }));
        expect(state).toMatchSnapshot();
      });
    });

    describe('apiRead()', () => {
      let state = fullStore;
      const apiWillRead = createAction(ActionTypes.API_WILL_READ);
      const apiReadDone = createAction(ActionTypes.API_READ_DONE);

      it('increments isReading', () => {
        state = reduce(undefined, apiWillRead({ data: person }));
        expect(state).toMatchSnapshot();
      });

      it('populates the state with resources', () => {
        state = reduce(state, apiReadDone({ data: person }));
        expect(state).toMatchSnapshot();
      });
    });

    describe('apiClear()', () => {
      it('clears the resource', () => {
        const state = reduce(fullStore, subject.apiClear(person.type));
        expect(state).toMatchSnapshot();
      });
    });

    describe('apiUpdate()', () => {
      let state = fullStore;
      const apiWillUpdate = createAction(ActionTypes.API_WILL_UPDATE);
      const apiUpdateDone = createAction(ActionTypes.API_UPDATE_DONE);

      it('increments isUpdating and marks the resource as pending', () => {
        state = reduce(state, apiWillUpdate({ data: person }));
        expect(state).toMatchSnapshot();
      });

      it('updates the state with resources', () => {
        state = reduce(state, apiUpdateDone({ data: person }));
        expect(state).toMatchSnapshot();
      });
    });

    describe('apiDelete()', () => {
      let state = fullStore;
      const apiWillDelete = createAction(ActionTypes.API_WILL_DELETE);
      const apiDeleteDone = createAction(ActionTypes.API_DELETE_DONE);

      it('increments isDeleting and marks the resource as pending', () => {
        state = reduce(state, apiWillDelete({ data: person }));
        expect(state).toMatchSnapshot();
      });

      it('removes the resources from the store', () => {
        state = reduce(state, apiDeleteDone({ data: person }));
        expect(state).toMatchSnapshot();
      });
    });

    describe('apiIndexClear()', () => {
      const hash = 'people.id=5,10';
      let state = {
        ...fullStore,
        indices: {
          [hash]: [{ id: person.id, type: person.type }],
        },
      };

      it('clears the index', async () => {
        state = reduce(state, subject.apiIndexClear(hash));
        expect(state.indices).toMatchSnapshot();
      });
    });

    describe('apiIndexSync()', () => {
      const hash = 'people.id=5,10';
      const store = mockStore;
      const { dispatch, getState } = store;
      const resources = { data: person };

      it('creates the index', () => {
        dispatch(subject.apiIndexSync({ hash, resources }));
        const state = getState().api.indices;
        expect(state).toMatchSnapshot();
      });
    });

    describe('apiIndexAsync()', () => {
      const hash = 'people.id=5,10';
      const store = mockStore;
      const { dispatch, getState } = store;

      beforeEach(() => {
        dispatch(subject.apiIndexClear(hash));
      });

      it('creates the index when the promise resolves', async () => {
        const promise = Promise.resolve({ data: person });
        await dispatch(subject.apiIndexAsync({ hash, promise }));
        const state = getState().api.indices;
        expect(state).toMatchSnapshot();
      });

      it('creates the index when the promise is rejected', async () => {
        const promise = Promise.reject({ data: person });
        try {
          await mockStore.dispatch(subject.apiIndexAsync({ hash, promise }));
        } catch (e) {
          // nothing
        }
        const state = mockStore.getState().api.indices;
        expect(state).toMatchSnapshot();
      });
    });
  });
});
