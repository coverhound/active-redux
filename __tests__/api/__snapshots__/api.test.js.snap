// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer actions apiClear() clears the resource 1`] = `
Object {
  "apiConfig": Object {},
  "indices": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {},
  },
}
`;

exports[`reducer actions apiConfigure() updates the apiConfig key 1`] = `
Object {
  "apiConfig": Object {
    "baseUrl": "http://coverhound.com/api",
  },
  "indices": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {},
}
`;

exports[`reducer actions apiCreate() increments isCreating 1`] = `
Object {
  "apiConfig": Object {},
  "indices": Object {},
  "isCreating": 1,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {},
}
`;

exports[`reducer actions apiCreate() populates the state with resources 1`] = `
Object {
  "apiConfig": Object {},
  "indices": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "5": Object {
        "attributes": Object {
          "age": 25,
          "hair": "black",
        },
        "id": 5,
        "type": "people",
      },
    },
  },
}
`;

exports[`reducer actions apiDelete() increments isDeleting and marks the resource as pending 1`] = `
Object {
  "apiConfig": Object {},
  "indices": Object {},
  "isCreating": 0,
  "isDeleting": 1,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "2": Object {
        "attributes": Object {
          "name": "Yangster",
        },
      },
      "5": Object {
        "attributes": Object {
          "name": "Johnny",
        },
        "isPending": true,
      },
    },
  },
}
`;

exports[`reducer actions apiDelete() removes the resources from the store 1`] = `
Object {
  "apiConfig": Object {},
  "indices": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "2": Object {
        "attributes": Object {
          "name": "Yangster",
        },
      },
    },
  },
}
`;

exports[`reducer actions apiHydrate() populates the state with resources 1`] = `
Object {
  "apiConfig": Object {},
  "indices": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "5": Object {
        "attributes": Object {
          "age": 25,
          "hair": "black",
        },
        "id": 5,
        "type": "people",
      },
    },
  },
}
`;

exports[`reducer actions apiIndexAsync() creates the index when the promise is rejected 1`] = `
Object {
  "people.id=5,10": Array [],
}
`;

exports[`reducer actions apiIndexAsync() creates the index when the promise resolves 1`] = `
Object {
  "people.id=5,10": Array [
    Object {
      "id": 5,
      "type": "people",
    },
  ],
}
`;

exports[`reducer actions apiIndexClear() clears the index 1`] = `
Object {
  "people.id=5,10": Array [],
}
`;

exports[`reducer actions apiIndexSync() creates the index 1`] = `
Object {
  "people.id=5,10": Array [
    Object {
      "id": 5,
      "type": "people",
    },
  ],
}
`;

exports[`reducer actions apiRead() increments isReading 1`] = `
Object {
  "apiConfig": Object {},
  "indices": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 1,
  "isUpdating": 0,
  "resources": Object {},
}
`;

exports[`reducer actions apiRead() populates the state with resources 1`] = `
Object {
  "apiConfig": Object {},
  "indices": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "5": Object {
        "attributes": Object {
          "age": 25,
          "hair": "black",
        },
        "id": 5,
        "type": "people",
      },
    },
  },
}
`;

exports[`reducer actions apiUpdate() increments isUpdating and marks the resource as pending 1`] = `
Object {
  "apiConfig": Object {},
  "indices": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 1,
  "resources": Object {
    "people": Object {
      "2": Object {
        "attributes": Object {
          "name": "Yangster",
        },
      },
      "5": Object {
        "attributes": Object {
          "name": "Johnny",
        },
        "isPending": true,
      },
    },
  },
}
`;

exports[`reducer actions apiUpdate() updates the state with resources 1`] = `
Object {
  "apiConfig": Object {},
  "indices": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "5": Object {
        "attributes": Object {
          "age": 25,
          "hair": "black",
        },
        "id": 5,
        "type": "people",
      },
    },
  },
}
`;

exports[`reducer initialState has an empty resources object 1`] = `
Object {
  "apiConfig": Object {},
  "indices": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {},
}
`;
